pyvista/themes.py

```diff
@@ -0,0 +1,1948 @@
+"""Module managing different plotting theme parameters."""
+
+import json
+from typing import Union
+
+import warnings
+from enum import Enum
+import os
+
+from .plotting.colors import PARAVIEW_BACKGROUND, get_cmap_safe
+from .plotting.tools import parse_color, parse_font_family
+from .utilities.misc import PyvistaDeprecationWarning
+
+
+def load_theme(filename):
+    """Load a theme from a file.
+
+    Examples
+    --------
+    >>> import pyvista
+    >>> theme = pyvista.themes.DefaultTheme()
+    >>> theme.save('my_theme.json')  # doctest:+SKIP
+    >>> loaded_theme = pyvista.load_theme('my_theme.json')  # doctest:+SKIP
+
+    """
+    return DefaultTheme.fromdict(json.load(open(filename)))
+
+
+def set_plot_theme(theme):
+    """Set the plotting parameters to a predefined theme."""
+    import pyvista
+    if isinstance(theme, str):
+        theme = theme.lower()
+        if theme == 'night':  # pragma: no cover
+            warnings.warn('use "dark" instead of "night" theme', PyvistaDeprecationWarning)
+        new_theme = ALLOWED_THEMES[theme].value()
+        pyvista.global_theme.load_theme(new_theme)
+    elif isinstance(theme, DefaultTheme):
+        pyvista.global_theme.load_theme(theme)
+    else:
+        raise TypeError(f'Expected a pyvista.Theme or str, not '
+                        f'a {type(theme)}')
+
+class _ThemeConfig():
+    """Provide common methods for theme configuration classes."""
+
+    @classmethod
+    def fromdict(cls, dict_):
+        """Create from a dictionary."""
+        inst = cls()
+        for key, value in dict_.items():
+            if not hasattr(inst, key):
+                raise KeyError(f'Invalid key "{key}" for {inst.__class__.__name__}')
+            attr = getattr(inst, key)
+            if hasattr(getattr(inst, key), 'fromdict'):
+                attr.fromdict(value)
```

Comment:
(originally posted on a commit rather than the PR)

I still don't want to go into rigorous code review as long as this is a draft, but this seems wrong here: you instantiate a new `_ThemeConfig` subclass, then discard it. You probably need something like
```
setattr(self, key, attr.fromdict(value))
```
(untested). The current implementation ignores changes to class-valued fields when doing something like
```py
theme = pv.themes.DefaultTheme()
theme.slider_style.classic.slider_length = 42
print(pv.themes.DefaultTheme.fromdict(theme._to_dict()) == theme)
# False
```
Assuming I understand correctly that this should be an idempotent roundtrip.

Reply:
Fixed in https://github.com/pyvista/pyvista/pull/1310/commits/c630fefe78de8fe92ebe9b94ad2e22ec93c6c511